state/txindex/kv/kv.go

```diff
@@ -42,6 +48,240 @@ type TxIndex struct {
+		if err != nil {
+			return fmt.Errorf("failed to flush tx indexer pruning batch %w", err)
+		}
+		err = batch.Close()
+		if err != nil {
+			txi.log.Error(fmt.Sprintf("Error when closing tx indexer pruning batch: %v", err))
+		}
+		return nil
+	}
+
+	sort.Sort(TxResultByHeight(results))
+	numHeightsBatchPruned := int64(0)                     // number of heights pruned if counting batched
+	currentBatchRetainedHeight := results[0].Height       // height retained if counting batched
+	numHeightsPersistentlyPruned := int64(0)              // number of heights pruned persistently
+	currentPersistentlyRetainedHeight := lastRetainHeight // height retained persistently
+	for i, result := range results {
+		errDeleteResult := txi.deleteResult(result, batch)
+		if errDeleteResult != nil {
+			// If we crashed in the middle of pruning the height,
+			// we assume this height is retained
+			errSetLastRetainHeight := txi.setIndexerRetainHeight(result.Height, batch)
+			if errSetLastRetainHeight != nil {
+				return 0, lastRetainHeight, fmt.Errorf("error setting last retain height '%v' while handling result deletion error '%v' for tx indexer", errSetLastRetainHeight, errDeleteResult)
+			}
+			errWriteBatch := batch.WriteSync()
+			if errWriteBatch != nil {
+				return 0, lastRetainHeight, fmt.Errorf("error writing tx indexer batch '%v' while handling result deletion error '%v'", errWriteBatch, errDeleteResult)
+			}
+			return result.Height - lastRetainHeight, result.Height, errDeleteResult
+		}
+		if i == len(results)-1 || results[i+1].Height > result.Height {
+			numHeightsBatchPruned++
+			currentBatchRetainedHeight = result.Height + 1
+		}
+		// flush every 1000 blocks to avoid batches becoming too large
+		if pruned%1000 == 0 && pruned > 0 {
+			err := flush(batch)
+			if err != nil {
+				return numHeightsPersistentlyPruned, currentPersistentlyRetainedHeight, err
+			}
+			numHeightsPersistentlyPruned = numHeightsBatchPruned
+			currentPersistentlyRetainedHeight = currentBatchRetainedHeight
+			batch = txi.store.NewBatch()
+			defer closeBatch(batch)
+		}
+	}
+
+	err = flush(batch)
+	if err != nil {
+		return numHeightsPersistentlyPruned, currentPersistentlyRetainedHeight, err
+	}
+	numHeightsPersistentlyPruned = numHeightsBatchPruned
+	currentPersistentlyRetainedHeight = currentBatchRetainedHeight
+	return numHeightsPersistentlyPruned, currentPersistentlyRetainedHeight, nil
+}
+
+func (txi *TxIndex) SetRetainHeight(retainHeight int64) error {
+	return txi.store.SetSync(TxIndexerRetainHeightKey, int64ToBytes(retainHeight))
+}
+
+func (txi *TxIndex) GetRetainHeight() (int64, error) {
+	buf, err := txi.store.Get(TxIndexerRetainHeightKey)
+	if err != nil {
+		return 0, err
+	}
+	if buf == nil {
+		return 0, state.ErrKeyNotFound
+	}
+	height := int64FromBytes(buf)
+
+	if height < 0 {
+		return 0, state.ErrInvalidHeightValue
+	}
+
+	return height, nil
+}
+
+func (txi *TxIndex) setIndexerRetainHeight(height int64, batch dbm.Batch) error {
```

Comment:
Consider renaming to `setLastRetainHeight`, to be consistent with the block indexer